### ISSUE TITLE
About your result output

### ISSUE THREAD

**Issue by wcy940418** [Vladkryvoruchko/PSPNet-Keras-tensorflow#2]
Hello, thank you for your excellent work. I am also interested in implementing PSPNet in TF. I have not tested your code yet, but about your output, I found some post process code in original matlab code.

<pre>
function score = ms_caffe_process(input_data,net)
    score = net.forward({input_data});
    score = score{1};
    score_flip = net.forward({input_data(end:-1:1,:,:)});
    score_flip = score_flip{1};
    score = score + score_flip(end:-1:1,:,:);

    score = permute(score, [2 1 3]);
    score = exp(score);
    score = bsxfun(@rdivide, score, sum(score, 3));
end
</pre>

It seems that they feed in the image twice each time, and one is flipped. Do you know why they did this?

Thank you

**Comment by Vladkryvoruchko** [comment]
Hi @wcy940418 . I havent found any implementation details about this processing in the original paper. As I dont have any experience in matlab, i can say that image is not actually flipped, but changed order of RGB channels to possible BGR... and then in score addition its changed back. In my pycaffe code i didnt do that "flipping" and it actually works pretty fine.

Will return to working with repo soon.

**Comment by wcy940418** [comment]
Thank you @Vladkryvoruchko , I see. I will try to figure out the missing detail in the original network.

**Comment by Vladkryvoruchko** [comment]
@wcy940418 Thank you. Keep me up to date if you proceed :)

**Comment by wcy940418** [comment]
@Vladkryvoruchko How did you modify the caffe-tensorflow code to port out bn layer parameters? And I tested using pycaffe to predict image, it just works fine as you said.

**Comment by wcy940418** [comment]
Also, I changed some lines in your code, now I get some better results, but not perfect. Could you tell me the order you convert scale, offset, mean, variance from caffemodel?

**Comment by Vladkryvoruchko** [comment]
@wcy940418 i will now upload in this repo original caffe-tensorflow code, and then commit modifications

**Comment by wcy940418** [comment]
Hi @Vladkryvoruchko , I found a snippet in your caffe-tensorflow code to port out bn params. In kaffe/transformers.py line 258:
<pre>
# Prescale the stats
  | scaling_factor = 1.0 / scale if scale != 0 else 0
  | mean *= scaling_factor
  | variance *= scaling_factor
</pre>
I am not sure mean and variance need to be prescaled. I think the meaning of four params of bn layer are identical in both caffe and tensorflow: scale * (data - mean) / variance + offset

**Comment by Vladkryvoruchko** [comment]
@wcy940418 Yea, but this snippet is from original converter and should work just fine. If we look on the tutorial of converting DilatedNet [link](http://nicolovaligi.com/converting-deep-learning-model-caffe-keras.html) there is also suppoed to be BN and it works fine there

**Comment by wcy940418** [comment]
Okay, I see. Look at here, in keras batchnormolization core function definition:
<pre>
        if self.scale:
            self.gamma = self.add_weight(shape=shape,
                                         name='gamma',
                                         initializer=self.gamma_initializer,
                                         regularizer=self.gamma_regularizer,
                                         constraint=self.gamma_constraint)
        else:
            self.gamma = None
        if self.center:
            self.beta = self.add_weight(shape=shape,
                                        name='beta',
                                        initializer=self.beta_initializer,
                                        regularizer=self.beta_regularizer,
                                        constraint=self.beta_constraint)
        else:
            self.beta = None
        self.moving_mean = self.add_weight(
            shape=shape,
            name='moving_mean',
            initializer=self.moving_mean_initializer,
            trainable=False)
        self.moving_variance = self.add_weight(
            shape=shape,
            name='moving_variance',
            initializer=self.moving_variance_initializer,
            trainable=False)
        self.built = True
</pre>
The order of parameters in bn layer is scale, offset, mean, variance.
But in your code:
<pre>
model.get_layer(layer.name).set_weights([mean, variance, scale, offset])
</pre>
The assigning order is mean, variance, scale, offset.
But even I change the assigning order, the result is not meaningful. So I still consider the ported parameters either have the wrong name or have unnecessary pre-process as I said before. Do you have any idea of this?

**Comment by wcy940418** [comment]
And I found another wrong code. Here in layers_builder.py line 177:
<pre>
	for i in range(3): #for i in range(2): old wrong code 
</pre>
level 3 should has 3_1 - 3_4 according to the original network.

**Comment by wcy940418** [comment]
I got even better result by finding out that there is a 
<pre>
prev_layer = Activation('relu')(prev_layer)
</pre>
lost at the top of 
<pre>
def residual_short(prev_layer, level, pad=1, lvl=1, sub_lvl=1, modify_stride=False):
</pre>
And also the epsilon setting of bn_layer has to be 1e-5, which is different from the default setting of Keras 1e-3.
Anyway, the result is much better than before, but still has many miss predictions now. I think is should be the difference between keras bn can caffe bn layer.

**Comment by Vladkryvoruchko** [comment]
@wcy940418 wow. I will now test it. impressive

**Comment by wcy940418** [comment]
@Vladkryvoruchko But I used Pillow(PIL) to read the image for my pycaffe code, the result is just fine as the original code.

**Comment by Vladkryvoruchko** [comment]
@wcy940418 Oh, I've missed such a small details, fixed by your modifications. Can you please make a pull request to this repo :)
